doc/sr_sara.1.rst

=========
 SR_Sara
=========

------------------------------------------
Subscribe, Acquire and ReAnnounce Products
------------------------------------------

:Manual section: 1 
:Date: Oct 2015
:Version: 0.0.1
:Manual group: MetPx Sarracenia Suite


SYNOPSIS
========

**sr_sara** configfile start|stop|restart|reload|status

DESCRIPTION
===========

**sr_sara** is a program that Subscribes to file notifications, 
Acquires the files and ReAnnounces them at their new locations.

The notification protocol is defined here `sr_post(7) <sr_post.7.html>`_

**sr_sara** connects to a *source_broker* (often the same as the remote file server 
itself) and subscribes to the notifications of interest. It uses the 
notification information to download the file on the local server its running on. 
After, it produces a new notification for the local file on a broker (usually on the local server).


**sr_sara** can be used to acquire files from `sr_post(1) <sr_post.1.html>`_
or `sr_watch(1) <sr_watch.1.html>`_  or to reproduce a web-accessible folders (WAF),
that announce its' products.

The **sr_sara** command takes two argument: a configuration file described below,
followed by an action start|stop|restart|reload|status... (self described).

CONFIGURATION
=============

Options are placed in the configuration file, one per line, of the form: 

**option <value>** 

Comment lines begins with **#**. 
Empty lines are skipped.


INSTANCES
---------

It is possible that one instance of sr_sara using a certain config
is not enough to process & download all available notifications.

**instances      <integer>     (default:1)**

sr_sara "configname" start   will fork  N instances of sr_sara using that config.
.sr_sara_configname_$instance.pid  are created and contain the PID  of $instance process.
sr_sara_configname_$instance.log  are created and contain the logs of $instance process.

The logs can be written in another directory than the current one with option :

**log            <directory logpath>  (default:$PWD)**


.. NOTE:: 
  FIXME: standard installation/setup explanations ...
  FIXME: a standard place where all configs ?
  FIXME: a standard place where all the logs ?
  FIXME: a standard place where all the pid files ?
  FIXME: a sr process starting all sara configs of various kind.



SOURCE NOTIFICATION OPTIONS
---------------------------

First, the program needs to set all the rabbitmq configurations for a source broker.
The source_broker option sets all the credential information to connect to the **AMQP** server 

**source_broker amqp{s}://<user>:<pw>@<brokerhost>[:port]/<vhost>**

::

      (default: amqp://guest:guest@localhost/ ) 


Once connected to an AMQP broker, the user needs to create a queue and bind it
to an exchange.  These options define which messages (URL notifications) the program receives:

 - **source_exchange      <name>         (default: amq.topic)** 
 - **source_topic         <amqp pattern> (default: v02.post.#)**
 - **queue_name           <name>         (default: sr_sara.config_name)** 

.. NOTE::
  FIXME: queue_name is sr_sara.config_name? is that the file that contains the queue name?
  FIXME: probably should be "queue_info_file" ? is there other stuff in there other than the name?
  FIXME: queue_name is supposed to be q_guest (assuming default auth.)

To give a correct value to the subtopic, browse the remote server and
write down v02.post plus the directory of interest separated by a dot
as follow:

 **source_topic  v02.post.directory1.*.subdirectory3.*.subdirectory5.#** 

::

 where:  
       *                replaces a directory name 
       #                stands for the remaining possibilities

One has the choice of filtering using  **source_topic**  with only AMQP's limited 
wildcarding. 

.. NOTE:: 
  FIXME: no mention of the use of exchange argument.
  FIXME: defaults should be wiser or None...
  FIXME: queue settings : no support in code default auto_delete false,durable false
  FIXME: one source_topic only : no support for a list in code
  FIXME: option source_subtopic,source_topic_prefix : no support in code


DESTINATION OPTIONS
-------------------

Theses options set where and how the program will place the files to be downloaded.

::

**document_root <path>           (default: .)** 
**mirror        <boolean>        (default: true)** 
**strip         <integer>        (default: 0)** 
**overwrite     <boolean>        (default: true)** 
**inplace       <boolean>        (default: true)** 

The **document_root** sets a directory the root of the download tree.
This directory never appears in the newly created amqp notifications.

By default, **mirror** option is True, the default path for a file is :

path = document_root + 'notification filepath'

**sr_sara** expects the notification filepath to start with YYYYMMDD/sourceid.
The user will set **mirror** to False, if it is not the case. The path
for the file becomes :

path = document_root + YYYYMMDD/sourceid + 'notification_filepath'

The **strip** option defines the number of directories to remove
from the path... This applies for subdirectories starting after the document_root
If the number of directories is greated than the subdirectories the path would
become :

path = document_root + filename


Once the path is defined in the program, if the **overwrite** option is set to True,'
the program checks if the file is already there. If it is, it computes the checksum
on it according to the notification'settings. If the local file checksum matches the
one of the notification, the file is not downloaded, the incoming notification is 
acknowledge, and the file is not announced. If the file is not there, or the checksum
differs, the file is overwritten and a new notification is sent to the destination broker.

The **inplace** option defaults to True. The program receiving notifications of file 
parts, will put these parts inplace in the file in an orderly fashion. Each parts,
once inserted in the file, is notified to the destination broker.


.. NOTE:: 
  - FIXME: lock option    : should it support file locking (.tmp, . prefix) ?
  - FIXME: destfn script  : should it support a destination script
  - FIXME: renamer script : should it support a file renamer script
  - FIXME: working_directory ? .. should this be a config option to name where the queue_name, and other? state files live?



DOWNLOAD CREDENTIALS 
--------------------

**ssh_keyfile  <filepath> (set if needed for sftp downloads)** 

.. NOTE::
  FIXME: usage of ~/.conf/sara/credentials.conf to be coded
  support of various ftp/sftp... etc credentials at the same time
  much easier for users and less restrictions on notifications


OUTPUT NOTIFICATION OPTIONS
---------------------------

The program needs to set all the rabbitmq configurations for an output broker.

The broker option sets all the credential information to connect to the
  output **RabbitMQ** server 

**broker amqp{s}://<user>:<pw>@<brokerhost>[:port]/<vhost>**

::

      (default: amqp://guest:guest@localhost/ ) 

Once connected to the source AMQP broker, the program builds notifications after
the download of a file has occured. To build the notification and send it to
the next hop broker, the user sets these options :

 - **url               <url>          (needs to be set)**
 - **recompute_chksum  <boolean>      (False)** 
 - **exchange          <name>         (default: amq.topic)** 

The **url** option sets how to get the file... it defines the protocol,
host, port, and optionally, the credentials. It is a good practice not to 
notify the credentials and separately inform the consumers about it.

If **recompute_chksum** is set to True, the checksum will be recomputed
on file download and value will overwrite the one from the incoming amqp message.  
If a file is repeatedly modified, the download may occur after the file is overwritten
but with its old notification... resulting in a checksum mismatch and potential
looping in a network of pumps.

.. NOTE::
   FIXME:  this is pathological case.  It ignores the incoming checksum.
   so data is forwarded in spite of checksum mismatch. We should think more about this.
   not sure this option is a good thing.


The **exchange** option set under which exchange the new notification will be posted.


QUALITY ASSURANCE
-----------------

These options can be used for quality assurance.

::

**message_validation_script    <script_path> (used if set)** 
**file_validation_script       <script_path> (used if set)** 

The  **message_validation_script**  receives a sr_message instance
containing all the amqp information. The user can write checks on
any of the sr_message values.  Should it not comply to the checks,
a log message (and an amqp log message) will posted, the message will be
acknowledged with out any further processing...  Only valid messages
will be processed further. 

.. NOTE:: 
  FIXME: where should we put these scripts
  FIXME: details missing in doc on returned values

The return values of this script are :
OK,code,message    <boolean,integer,string>   accepted?,error code, error message


The  **file_validation_script**  receives the file path.
The user may run any kind of validation on the path.
Should the file not comply to the checks, a log message (and an amqp log message) will posted,
the message will be acknowledged without any further processing... 
Only valid files are reannounced.

.. NOTE:: 
  FIXME: where should we put these scripts
  FIXME: what should we do with rejected files ... validation script removes it ?
  FIXME: details missing in doc on returned values

The return values of this script are :
OK,code,message    <boolean,integer,string>   accepted?,error code, error message



.. NOTE:: 
  FIXME: accept/reject should be coded... and documented
  not sure if we need accept/reject... interesting...
  work on other stuff first...

SEE ALSO
========

`sr_log(7) <sr_log.7.html>`_ - the format of log messages.

`sr_post(1) <sr_post.1.html>`_ - post announcemensts of specific files.

`sr_post(7) <sr_post.7.html>`_ - The format of announcements.

`sr_subscribe(1) <sr_subscribe.1.html>`_ - the download client.

`sr_watch(1) <sr_watch.1.html>`_ - the directory watching daemon.

`dd_subscribe(1) <dd_subscribe.1.html>`_ - the http-only download client.


sara/sr_sara.py
#!/usr/bin/python3
#
# This file is part of sarracenia.
# The sarracenia suite is Free and is proudly provided by the Government of Canada
# Copyright (C) Her Majesty The Queen in Right of Canada, Environment Canada, 2008-2015
#
# Questions or bugs report: dps-client@ec.gc.ca
# sarracenia repository: git://git.code.sf.net/p/metpx/git
# Documentation: http://metpx.sourceforge.net/#SaraDocumentation
#
# sr_sara.py : python3 program allowing users to listen and download product from
#              another sarracenia server or from user posting (sr_post/sr_watch)
#
#
# Code contributed by:
#  Michel Grenier - Shared Services Canada
#  Jun Hu         - Shared Services Canada
#  Last Changed   : Sep 22 10:41:32 EDT 2015
#  Last Revision  : Sep 22 10:41:32 EDT 2015
#
########################################################################
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful, 
#  but WITHOUT ANY WARRANTY; without even the implied warranty of 
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
#
#

import os,sys,time

try :    
         from sr_amqp           import *
         from sr_file           import *
         from sr_ftp            import *
         from sr_http           import *
         from sr_instances      import *
         from sr_message        import *
         from sr_sftp           import *
except : 
         from sara.sr_amqp      import *
         from sara.sr_file      import *
         from sara.sr_ftp       import *
         from sara.sr_http      import *
         from sara.sr_instances import *
         from sara.sr_message   import *
         from sara.sr_sftp      import *

class sr_sara(sr_instances):

    def __init__(self,config=None,args=None):
        sr_instances.__init__(self,config,args)
        self.defaults()
        self.exchange = 'xpublic'
        self.configure()

    def check(self):

        # dont want to recreate these if they exists

        if not hasattr(self,'msg') :
           self.msg      = sr_message(self.logger)

        self.msg.user = self.source_broker.username

        # expand self.from_cluster in case it is an alias

        self.from_cluster_list = []
        if self.from_cluster != None :
           self.from_cluster_list.append(self.from_cluster)
           if self.from_cluster in self.aliases :
              self.from_cluster_list.extend(self.aliases[self.from_cluster])

        # umask change for directory creation and chmod

        try    : os.umask(0)
        except : pass

    def close(self):
        self.hc_src.close()
        self.hc_pst.close()

    def connect(self):

        # =============
        # consumer
        # =============

        # consumer host

        self.hc_src = HostConnect( logger = self.logger )
        self.hc_src.set_url( self.source_broker )
        self.hc_src.connect()

        # consumer  add_prefetch(1) allows queue sharing between instances

        self.consumer  = Consumer(self.hc_src)
        self.consumer.add_prefetch(1)
        self.consumer.build()

        # consumer queue
        name  = 'q_' + self.source_broker.username + '.' + self.program_name + '.' + self.config_name
        if self.queue_name != None :
           name = 'q_' + self.source_broker.username + '.' + self.queue_name

        self.queue = Queue(self.hc_src,name)
        self.queue.add_binding(self.source_exchange,self.source_topic)
        self.queue.build()

        # log publisher

        self.amqp_log    = Publisher(self.hc_src)
        self.amqp_log.build()

        # =============
        # publisher
        # =============

        # publisher host

        self.hc_pst = HostConnect( logger = self.logger )
        self.hc_pst.set_url( self.broker )
        self.hc_pst.connect()

        # publisher

        self.amqp_pub    = Publisher(self.hc_pst)
        self.amqp_pub.build()

    def configure(self):

        # installation general configurations and settings

        self.general()

        # arguments from command line

        self.args(self.user_args)

        # config from file

        self.config(self.user_config)

        # verify all settings

        self.check()

        self.setlog()
        self.logger.info("user_config = %d %s" % (self.instance,self.user_config))

    def delete_event(self):
        if self.msg.sumflg != 'R' : return False

        self.msg.code = 503
        self.msg.message = "Service unavailable : delete"
        self.msg.log_error()
 
        return True


    def download(self):

        self.logger.info("downloading/copying into %s " % self.msg.local_file)

        try :
                if   self.msg.url.scheme == 'http' :
                     return http_download(self.msg, self.http_user, self.http_password )

                elif self.msg.url.scheme == 'ftp' :
                     return ftp_download(self.msg, self.ftp_user, self.ftp_password, self.ftp_mode, self.ftp_binary )

                elif self.msg.url.scheme == 'sftp' :
                     return sftp_download(self.msg, self.sftp_user, self.sftp_password, self.sftp_keyfile )

                elif self.msg.url.scheme == 'file' :
                     return file_process(self.msg)

        except :
                (stype, svalue, tb) = sys.exc_info()
                self.logger.error("Download  Type: %s, Value: %s,  ..." % (stype, svalue))
                self.msg.code    = 503
                self.msg.message = "Unable to process"
                self.msg.log_error()

        self.msg.code    = 503
        self.msg.message = "Service unavailable %s" % self.msg.url.scheme
        self.msg.log_error()


    def help(self):
        self.logger.info("Usage: %s [OPTIONS] configfile [start|stop|restart|reload|status]\n" % self.program_name )
        self.logger.info("OPTIONS:")
        self.logger.info("-i   <nbr_of_instances>      default 1")
        self.logger.info("-b   <broker>                default amqp://guest:guest@localhost/")
        self.logger.info("-dr  <document_root>")
        self.logger.info("-ex  <exchange>              default xpublic")
        self.logger.info("-u   <url>")
        self.logger.info("-sb  <source_broker>         default amqp://guest:guest@localhost/")
        self.logger.info("-se  <source_exchange>       default xpublic")
        self.logger.info("-st  <source_topic>          default v02.post.#")
        self.logger.info("-qn  <queue_name>            default q_username.config_name")
        self.logger.info("-m   <mirror>                default True")
        self.logger.info("-ms  <message_validation_script>")
        self.logger.info("-fs  <file_validation_script>")
        self.logger.info("-sk  <ssh_keyfile>")
        self.logger.info("-strip <strip count (directory)>")
        self.logger.info("-in  <put parts inplace>      default True")
        self.logger.info("-o   <overwrite no sum check> default True")
        self.logger.info("-rc  <recompute sun>          default False")
        self.logger.info("DEBUG:")
        self.logger.info("-debug")

    # this sr_sara runs for cluster : self.from_cluster
    # this sr_sara is a gateway for a list of clusters : self.route 
    # the self.route list, is expanded... so contains aliases and their defined clusters 
    # A message has a list of targetted clusters : self.msg.to_list
    # There is a list of defined cluster aliases : self.aliases
    # A message's target can be an alias.

    # From all this info, this module decides if sr_sara should
    # treat that message (return True)  or not (return False)

    def route_this_message(self):

        # the message has not specified a destination.
        if not 'to' in self.msg.headers :
           self.msg.code    = 403
           self.msg.message = "Forbidden : message without destination amqp header['to']"
           self.msg.log_error()
           return False

        # loop on all message destinations (target)
        for target in self.msg.to_list :

           # target is in this cluster
           if target in self.from_cluster_list : return True

           # target is in this cluster's defined route
           if target in self.route : return True

           # if target is not an alias ... we are done with it
           if not target in self.aliases : continue

           # loop on all alias names for this target

           for target2 in self.aliases[target] :

               # target2 is in this cluster
               if target2 in self.from_cluster_list : return True

               # target2 is in this cluster's defined route
               if target2 in self.route : return True

        # nope this one is not for this cluster
        self.logger.warning("skipped : not for this cluster...")

        return False

    def set_local(self):

        # relative path by default mirror 

        self.rel_path = '%s' % self.msg.path
        if 'rename' in self.msg.headers :
           self.rel_path = '%s' % self.msg.headers['rename']

        # if we dont mirror force  yyyymmdd/source in front

        if not self.mirror :
           yyyymmdd = time.strftime("%Y%m%d",time.gmtime())
           self.rel_path = '%s/%s/%s' % (yyyymmdd,self.msg.headers['source'],self.msg.path)

           if 'rename' in self.msg.headers :
              self.rel_path = '%s/%s/%s' % (yyyymmdd,self.msg.headers['source'],self.msg.headers['rename'])
              self.rel_path = self.rel_path.replace('//','/')

        token = self.rel_path.split('/')
        self.filename = token[-1]

        # if strip is used... strip N heading directories

        if self.strip > 0 :
           if self.strip > len(token) : token = [token[-1]]
           else :                       token = token[self.strip:]
           self.rel_path = '/'.join(token)


        self.local_dir  = ''
        if self.document_root != None :
           self.local_dir = self.document_root 
        if len(token) > 1 :
           self.local_dir = self.local_dir + '/' + '/'.join(token[:-1])

        # Local directory (directory created if needed)

        self.local_dir  = self.local_dir.replace('//','/')
        self.local_file = self.local_dir   + '/' + self.filename
        self.local_url  = urllib.parse.urlparse(self.url.geturl() + '/' + self.rel_path)

        # we dont propagate renaming... once used get rid of it
        if 'rename' in self.msg.headers : del self.msg.headers['rename']

    def validate_file(self,target_file, offset, length):

        if self.file_script == None : return True

        ok, code, message = self.file_script(target_file, offset, length)

        if not ok :
           self.msg.code    = code
           self.msg.message = message
           self.msg.log_error()

        return ok

    def validate_message(self):

        if self.msg_script == None : return True

        ok, code, message = self.msg_script(self.msg)

        if not ok :
           self.msg.code    = code
           self.msg.message = message
           self.msg.log_error()

        return ok


    def run(self):

        self.logger.info("sr_sara run")

        self.connect()

        self.msg.logger       = self.logger
        self.msg.amqp_log     = self.amqp_log
        self.msg.amqp_pub     = self.amqp_pub
        self.msg.exchange_pub = self.exchange

        #
        # loop on all message
        #

        raw_msg = None

        while True :

          try  :
                 if raw_msg != None : self.consumer.ack(raw_msg)

                 raw_msg = self.consumer.consume(self.queue.qname)
                 if raw_msg == None : continue

                 # make use it as a sr_message

                 self.msg.from_amqplib(raw_msg)
                 self.logger.info("Received %s '%s' %s" % (self.msg.topic,self.msg.notice,self.msg.hdrstr))

                 # if message for this cluster or for this cluster's route

                 ok = self.route_this_message()
                 if not ok : continue

                 # setting source from exchange 

                 if self.source_from_exchange :
                    ok = self.set_source()
                    if not ok : continue

                 # setting originating cluster if not defined

                 if not 'from_cluster' in self.msg.headers :
                    ok = self.set_from_cluster()
                    if not ok : continue

                 # message validation

                 if not self.validate_message() : continue

                 # set local file according to sara : dr + imsg.path (or renamed path)

                 self.set_local()

                 # set local file according to the message and sara's setting

                 self.msg.set_local(self.inplace,self.local_file,self.local_url)

                 # asked to delete ?

                 if self.delete_event() : continue

                 # make sure local directory exists
                 # FIXME : logging errors...
                 try    : os.makedirs(self.local_dir,0o775,True)
                 except : pass

                 # if overwrite is not enforced (False)
                 # verify if msg checksum and local_file checksum match
                 # FIXME : should we republish / repost ???

                 if not self.overwrite and self.msg.checksum_match() :

                    self.msg.code    = 304
                    self.msg.message = 'not modified'
                    self.msg.log_info()
             
                    # a part unmodified can make a difference
                    if self.inplace and self.msg.in_partfile :
                       file_reassemble(self.msg)

                    # chksum computed on msg offset/length may need to truncate
                    file_truncate(self.msg)
                    continue

                 # proceed to download  3 attempts

                 i  = 0
                 while i < 3 : 
                       ok = self.download()
                       if ok : break
                       i = i + 1
                 if not ok : continue

                 # validate file/data

                 if not self.validate_file(self.msg.local_file, self.msg.local_offset, self.msg.length) : continue

                 # force recompute checksum

                 if self.recompute_chksum :
                    self.msg.compute_local_checksum()

                    # When downloaded, it changed from the message... 
                    # this is weird... means the file changed in the mean time
                    # and probably reposted...
                    # now posting of this file will have accurate checksum

                    if not self.msg.local_checksum == self.msg.checksum :
                       self.msg.set_sum(self.msg.sumflg,self.msg.local_checksum)
                       self.msg.code    = 205
                       self.msg.message = 'Reset Content : checksum'
                       self.msg.log_info()


                 # Delayed insertion
                 # try reassemble the file, conditions may have changed since writing

                 if self.inplace and self.msg.in_partfile :
                    self.msg.code    = 307
                    self.msg.message = 'Temporary Redirect'
                    self.msg.log_info()
                    file_reassemble(self.msg)
                    continue

                 # announcing the download or insert

                 if self.msg.partflg != '1' :
                    if self.inplace : self.msg.change_partflg('i')
                    else            : self.msg.change_partflg('p')

                 self.msg.set_topic_url('v02.post',self.msg.local_url)
                 self.msg.set_notice(self.msg.local_url,self.msg.time)
                 self.msg.code    = 201
                 self.msg.message = 'Published'
                 self.msg.publish()
              
                 # if we inserted a part in file ... try reassemble

                 if self.inplace and self.msg.partflg != '1' :
                    file_reassemble(self.msg)

          except :
                 (stype, svalue, tb) = sys.exc_info()
                 self.logger.error("Type: %s, Value: %s,  ..." % (stype, svalue))

    def set_from_cluster(self):
        if self.from_cluster == None :
           self.msg.code    = 403
           self.msg.message = "Forbidden : message without cluster"
           self.msg.log_error()
           return False

        self.msg.set_from_cluster(self.from_cluster)
        return True

    def set_source(self):
        if self.msg.exchange[:3] != 'xs_' :
           self.msg.code    = 403
           self.msg.message = "Forbidden : message without source"
           self.msg.log_error()
           return False

        source = self.msg.exchange[3:]
        self.msg.set_source(source)
        return True

    def reload(self):
        self.logger.info("sr_sara reload")
        self.close()
        self.configure()
        self.run()

    def start(self):
        self.configure()
        self.logger.info("sr_sara start")
        self.run()

    def stop(self):
        self.logger.info("sr_sara stop")
        self.close()
        os._exit(0)


# ===================================
# MAIN
# ===================================

def main():

    action = None
    args   = None
    config = None

    if len(sys.argv) >= 3 :
       action = sys.argv[-1]
       config = sys.argv[-2]
       if len(sys.argv) > 3: args = sys.argv[1:-2]

    sara = sr_sara(config,args)

    if   action == 'reload' : sara.reload_parent()
    elif action == 'restart': sara.restart_parent()
    elif action == 'start'  : sara.start_parent()
    elif action == 'stop'   : sara.stop_parent()
    elif action == 'status' : sara.status_parent()
    else :
           sara.logger.error("action unknown %s" % action)
           sys.exit(1)

    sys.exit(0)



# =========================================
# direct invocation
# =========================================

if __name__=="__main__":
   main()
